openedx/frontend-app-communications
# feat: slot component for dynamic plugins

### Description

Following this [pluggable approach to importing plugins](https://github.com/openedx/frontend-app-course-authoring/pull/638), we've created a reusable component that will dynamically import a module wherever it is called. The main idea is to be able to change the plugins with this mechanism, which means that instead of modifying the micro-frontend beforehand, we could change the plugin in isolation.

![slot-compoentn](https://github.com/openedx/frontend-app-communications/assets/134975835/ad1cb179-320c-4e2f-a1c4-f9f2d00ac51c)





#### How does it work?

Any component can be wrapped with the pluggable component. For instance:

```jsx
<PluggableComponent id="input-form" as="communications-app-card">
    <h1>Any component as default</h1>
</PluggableComponent>
```

The prop `as` indicates where is the plugin allocated. In this case, the component will be pointing to this route: `node_modules/@openedx-plugins/communications-app-card`.

If the plugin was installed as a dependency, it will render the component that you have pointing to that one.

If the plugin isn't installed, it will return its children. In this case, it's this:
```jsx
<h1>Any component as default</h1>
```

#### Why are we doing this?

This means that the plugin doesn't have to be installed if I'm making a feature. It's something that we can improve in the future. That way, another developer can replace that component and make more changes. This approach allows us to wrap as many components as we can, making most of the features pluggable and adaptable to various use cases. The component can be rendered without children as well, which means it won't return anything if the plugin is not installed.


### How to test it

Check it out this branch and follow these steps:

```
 cd plugins/communications-app/CheckBoxForm && npm link
```
```
 cd plugins/communications-app/InputForm && npm link
```
```
 nvm use && npm install
```
Then you can run the mfe with `npm start`

If you want to create another plugin you can do it in the folder plugins/communications-app/ with the structure of the other
plugins
